execution/shuffle.go

```diff
@@ -0,0 +1,449 @@
+package execution
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/pkg/errors"
+
+	"github.com/cube2222/octosql"
+	"github.com/cube2222/octosql/streaming/storage"
+)
+
+func GetAndStartAllShuffles(ctx context.Context, stateStorage storage.Storage, tx storage.StateTransaction, nodes []Node, variables octosql.Variables) ([]RecordStream, []*ExecutionOutput, error) {
+	ctx = storage.InjectStateTransaction(ctx, tx)
+	rootShuffleID := &ShuffleID{Id: "root_shuffle_id"}
+	nextShuffleIDRaw, err := GetSourceStreamID(tx.WithPrefix(rootShuffleID.AsPrefix()), octosql.MakeString("next_shuffle_id"))
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get next shuffle id")
+	}
+	nextShuffleID := (*ShuffleID)(nextShuffleIDRaw)
+
+	nextShuffles := make(map[string]ShuffleData)
+
+	// We start the head node to possibly get the first shuffle
```

Comment:
I don't really understand what's going on here and I think it could use some more commenting (I mean the `for partition, node := range nodes` loop)
You go over all nodes you are trying to start, you get their recordStreams and executionOutputs from Get. What is `executionOutput.NextShuffles` for example? From what I understood you basically get all the shuffles from the nodes?

Reply:
if you have output <- Shuffle A <- Shuffle B <- source

then shuffle A has to generate the ID for shuffle B, it's the only way to make sure all input streams to A end up with the same B shuffle ID.